fix: Go - avoid panic on wrong *keyset.Handle

Baha-sk opened the pull request:
For each of the available primitives in Go - ensure type checking of the primitive to avoid panics in the API. 

closes #336

Signed-off-by: Baha Shaaban <baha.shaaban@securekey.com>

Baha-sk commented:
@chuckx , @kste , @happyCoder92, @juergw 
I noticed that 6 of your commits were merged today (not as PRs)

Can you please review this PR? I rebased on top of your commits.

tholenst commented:
Sorry for the delay.

We might change that code in the near future, so feel free to file an issue. Anyhow, I will comment on it.

tholenst commented on go/aead/aead_factory.go:
```diff
@@ -47,24 +48,34 @@ type wrappedAead struct {
 
-func newWrappedAead(ps *primitiveset.PrimitiveSet) *wrappedAead {
+func newWrappedAead(ps *primitiveset.PrimitiveSet) (*wrappedAead, error) {
+	if _, ok := (ps.Primary.Primitive).(tink.AEAD); !ok {
```
We should check all primitives in the keyset; they all need to be convertible to AEAD objects (otherwise, a keyset with mixed keys will silently behave weirdly, which is not the desired behavior).

Hence, we should check whether all primitives can be cast to AEAD here.

Baha-sk replied:
Sure, I will add the check on all primitives in the keyset.

Baha-sk replied:
done

Baha-sk commented:
> Can you change the construction of the wrappers such that they check the correctness on construction time? Thanks!

Do you mean remove the checks from the primitive functions? ie Encrypt(), Decrypt(), Sign(), etc.
I will do this as well.

tholenst commented:
> Do you mean remove the checks from the primitive functions? ie Encrypt(), Decrypt(), Sign(), etc.
> I will do this as well.

That's a style question. In principle, if these checks trigger there's a bug in our code. In other languages that cannot happen, so I don't have a definite answer. I guess I personally would leave them in, as crashing is also bad when we have a bug in our code, but I'd return an error if any of the errors triggers (instead of swallowing).

Baha-sk commented:
I updated my previous comment :)

Baha-sk commented:
> I guess I personally would leave them in, as crashing is also bad when we have a bug in our code, but I'd return an error if any of the errors triggers (instead of swallowing).


I think what you mean is return an error instead of `continue`  in the below condition?
```
if !ok {
 	continue
 }
```

My thinking is if the primitive is not valid, keep looping through the list until we get a valid one. Technically this will never happen because of the primitive check in the constructor especially with the update I will be pushing soon to verify all primitive entries. 

But if you think that returning an error here is better than `continue`, I can do that too.

Baha-sk commented:
@tholenst I've updated the PR by first rebasing it and replaced every `continue` added with a `return fmt.Errorf()` statement instead.

I believe the PR is ready.

tholenst commented:
Ty, working on it.

Baha-sk commented:
Not sure why PRs don't get merged directly in Github. But this PR has been merged internally by Google. 

So I'm closing this PR. 

Thanks @tholenst and @sophieschmieg

tholenst commented:
Because we have two repositories, the internal one and the external one. We need to decide which repository mirrors the other one, and in Tink github mirrors the Google internal one.

It seems in this case I didn't follow the process carefully though, and now the association is not proper, and points to me instead of you. If you like, I can see how to fix that (e.g by rolling it back and re-merging this PR, or maybe it's possible to edit the history in git). Let me know what you think.

Baha-sk commented:
> Because we have two repositories, the internal one and the external one. We need to decide which repository mirrors the other one, and in Tink the github repo mirrors the Google internal one.
> 
> It seems in this case I didn't follow the process carefully though, and now the association is not proper, and points to me instead of you. If you like, I can see how to fix that (e.g by rolling it back and re-merging this PR, or maybe it's possible to edit the history in git). Let me know what you think.

That's ok @tholenst, what matters is it's merged. There's no need to put extra effort in reverting the commits. 

In the future though, it would be nice to be able to port GitHub merged commits to Google's internal repo for sake of the open source spirit of the project. It should be at least automated to save time and effort.
